addon/corpus.js

```diff
@@ -1,106 +1,231 @@
-async function freezeAllPages() {
-    // Grey out Freeze button:
-    document.getElementById('freeze').disabled = true;
+let gUrls;  // Array of [filename, url] to freeze.
+let gUrlIndex;  // Pointer to current url in gUrls.
+let gFreezeOptions; // Freeze-dry options from UI.
+let gTimeout; // Load+Freeze timeout from UI.
+
+async function freezeAllPages(event) {
+    event.preventDefault();
+
+    // Check form validity.
+    if (!(
+        document.getElementById('wait').validity.valid &&
+        document.getElementById('timeout').validity.valid
+    )) {
+        return;
+    }
+
+    initialiseFromForm();
+    emptyElement(document.getElementById('status'));
 
-    // Clear error field:
-    const errorField = document.getElementById('errors');
-    while (errorField.firstChild) {
-        errorField.removeChild(errorField.firstChild);
+    // We need at least one url.
+    if (!gUrls.length) {
+        let li = document.createElement('li');
+        li.classList.add('error');
+        li.appendChild(document.createTextNode('no pages to download'));
+        document.getElementById('status').appendChild(li);
+        return;
     }
 
-    // Make freezing window, and set its size. The blank page acts as a
+    document.getElementById('freeze').disabled = true;
+    let windowId = 0;
+
+    // Listen for tab events before we start creating tabs; this avoids race
+    // conditions between creating tabs and waiting for loading to complete.
+    browser.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
+        if (
+            tab.windowId !== windowId ||
+            tab.url === 'about:blank' ||
+            tab.status !== 'complete'
+        ) {
+            return;
+        }
+
+        if (tab.url.startsWith('moz-extension://')) {
+            // The blank placeholder page has loaded, start the freezing process.
+            document.dispatchEvent(new CustomEvent(
+                'fathom:next',
+                { detail: windowId }
+            ));
+
+        } else {
+            // Tab that needs to be frozen has loaded.
+
+            // Setup the timeout; this covers both the page load and freeze time.
+            let timer = setTimeout(() => {
+                // Timeout.
+                console.error(tab.url, 'timeout');
+                clearTimeout(timer);
+                setCurrentStatus('timeout', true);
+
+                // Close the tab and process the next url.
+                browser.tabs.remove(tab.id)
+                    .then(() => {
```

Comment:
This is one of those spots where I feel like .then() just adds noise. I think it's that no params are passed, so there wouldn't need to be any extra vars instantiated in the `await` equivalent.

Reply:
i suspect we'll never agree on `then` vs `async` 😃
will change to `async` here.

Reply:
Someday I hope to abandon the whole language. :-)